There are situations where built-in functions are not available in a library. In such cases, we do the computations by using formulas/concepts

At times there are no built-in functions available, thus there arises a need for the user to define a function as desired. The user-defined functions are named blocks of code that are designed to do a specific job(s). When you want to perform a particular task that you’ve defined in a function, you call the name of the function responsible for it. This aids the scenarios where a chunk of code is required repeatedly; just call the function dedicated to handling that task, and the call tells Python to run the code inside the function. You’ll find that using functions makes your programs easier to write, read, test, and fix.

### Argument and Parameter

Parameters and arguments are subtly different. Let us first define them.<br><br>
1. Parameter: A value which is passed when defining a function
2. Argument: A value that is passed when invoking a function




def fun(params):
    statement 1
    statement 2
        .
        .
        .

Now, if we invoke this function:

Here the 'sample_parameters' is called an argument.

### User defined functions without arguments

Once a function is defined, it can be called anytime, without running the function again. Generally, a function returns a value or prints a value.  

Let us see a few examples of user defined functions.

**1. We shall write a user defined function without argument to find the twice of a number.**

In [10]:
def twice():
    a = int(input('enter a number: '))
    print(a*2)

In [11]:
# In this function, no argument is passed
# call the function
twice()

4


**2. We shall write a user defined function hi() which takes a string input. The output will greet the user displaying his/her name on the screen.**

In [8]:
def hi():
    name = str(input('Enter your name: '))
    print("Hey " + name +", how are you?")

In [12]:
# In this function, no argument is passed
# call the function
hi()

Hey Amaan, how are you?


In [6]:
#create a function
def greetings():
    print("hello")
    print("How are you ?")

In [15]:
# In this function, no argument is passed
# call the function
greetings()

NameError: name 'greetings' is not defined

**3. To find area of a circle**

In [17]:
def area_circle():
    r = int(input('Enter the radius of circle: '))
    area = 3.14 * r**2 
    print("The area of circle with radius " + str(r)  +" is " + str(area) + " sq. units")

area_circle()

The area of circle with radius 45 is 6358.5 sq. units


calling function without argument in diffrent cell would not work


In [16]:
# call the function   
area_circle()

NameError: name 'area_circle' is not defined

### 2.2 User defined functions with arguments

Most of the functions will either `return` a value or `print` an output. Take a note of this in the examples considered below.

**1. A user defined function to get the average of all numbers in a sequence.**

In [10]:
def average(x):
    count = 0
    total = 0
    for i in x:
        count += 1
        total += i
    avg = total/count
    return avg

In [11]:
x = (44, 45, -56, 44, 32,-54, 44, 56, 44, 87, 44, 56)

# call the function   
average(x)

32.166666666666664

The average of the number is 32.1667

If the function requires an argument to be passed and we do not provide it, then an error will be shown.

In [15]:
# call the function   
average()

TypeError: average() missing 1 required positional argument: 'x'

**2. A user defined function with argument to greet a person by name.**

In [16]:
def greet(name):
    name = str(name)
    print("Hello, " + name)

In [17]:
# call the function        
greet('Dave')

Hello, Dave


Enter your name in the parenthese, and your notebook will greet you!.

**Also note that this function prints an output.** 

**Answer:** The 'print' output can not be stored in a variable. Moreover, any function which returns a value can be stored in a variable. 

**Consider the list given below and write a function to find the unique values of the list**

                [44,45,56,44,32,54,44,56,44,87,44,56]

In [18]:
def unique(list1): 
    unique_list = [] 
    for x in list1: 
            if x not in unique_list: 
                 unique_list.append(x)  
    return unique_list    

In [19]:
x = [44,45,56,44,32,54,44,56,44,87,44,56]

# call the function  
# store the output in a variable
unique_values = unique(x)

# print the variable name
unique_values

[44, 45, 56, 32, 54, 87]

These are the unique values in the list.

**Also note that this function returns the output to the calling program.**

**A function to reverse a string**

In [20]:
def reverse(string):
    # set the empty string
    rev=''
    for i in string:
        rev = i+rev
    print(rev)

In [22]:
# call the function
string = str(input('Enter string:'))
reverse(string)

Enter string:Data Science
ecneicS ataD


**A function to display multiplication table of an integer**

In [23]:
def table(n):
    multi = 1
    for i in range(1,11):
        multi = n*i
        print(str(i) + " times " + str(n) + " is " + str(multi))

In [24]:
# call the function        
n=int(input('Enter number:'))
table(n)

Enter number:6
1 times 6 is 6
2 times 6 is 12
3 times 6 is 18
4 times 6 is 24
5 times 6 is 30
6 times 6 is 36
7 times 6 is 42
8 times 6 is 48
9 times 6 is 54
10 times 6 is 60


**To find the product of numbers in a list.**

In [25]:
def product(x):
    prod = 1
    for i in x:
        prod *= i
    return prod

In [26]:
#call the function
product([1,2,3,4,5])

120

**A function which takes the order of a matrix and multiplies it with a scalar. The function takes two parameters- first the order of identity matrix and second the scalar to be multiplied.**

In [27]:
def scalar_multiplication(n,m):
    for i in range(0,n):
        for j in range(0,n):
            if(i==j):
                print(m,sep=" ",end=" ")
            else:
                print("0",sep=" ",end=" ")
        print()

In [28]:
#call the function
scalar_multiplication(4,5)

5 0 0 0 
0 5 0 0 
0 0 5 0 
0 0 0 5 


**Usage of `*args`.**

Usage : multiple arguments can be added without specifying the number of parameters defined in a function.  It is used to pass a non-keyworded, variable-length argument list.

 `*args` in function definitions in python is used to pass a variable number of arguments to a function. It reads the value one by one and performs the required task. <br>
 The symbol `*` is used to take in a variable number of arguments; by convention, it is often used with the word 'args', however we may use another word.

In [29]:
# * indicates that number of arguments can be taken
def Multiple(*args):  
    for parameter in args:  
        print (parameter) 

In [30]:
#call the function
Multiple("Lets","learn","Python","for","Data","Science") 

Lets
learn
Python
for
Data
Science


**Another example on `*args`.**

In [31]:
def to_print( argument1, *argument):
   print(argument1)
   for var in argument:
        print(var)

In [32]:
#call the function        
to_print( 10,20,30 )   

10
20
30


**Usage of  ` **kwargs`.**

Usage: `**kwargs` is used to pass a keyworded, variable-length argument list to the function. The name 'kwargs' is used with the double star because it allows us to pass through keyword arguments (and any number of them).

In [33]:
def function(**kwargs):
    for a,b in kwargs.items():
        print(a,"==",b)

In [34]:
# call the function
a=function(A = "First letter", B = "Second letter", C = "Third letter")

A == First letter
B == Second letter
C == Third letter


**Another example of ` **kwargs `.**

`**kwargs` is a dict of the keyword or arguments passed to the function

In [35]:
def print_keyword_args(**kwargs):
      for key, value in kwargs.items():
            print( "%s = %s" % (key, value))

In [36]:
# call the function
print_keyword_args(first_name="John", last_name="Doe")

first_name = John
last_name = Doe
